tmt/steps/provision/__init__.py

```diff
@@ -581,68 +581,55 @@ def _query_distro(self, guest: 'Guest') -> Optional[str]:
-        # one available. Collect them, and sort them by their priorities
-        # to find the most suitable one.
+        # Sort available package managers by priority and probe them one by one,
+        # break after the first one is detected.
 
         discovered_package_managers: list[
             PackageManagerClass[tmt.package_managers.PackageManagerEngine]
-        ] = [
-            package_manager_class
-            for package_manager_id, package_manager_class in tmt.package_managers._PACKAGE_MANAGER_PLUGIN_REGISTRY.items()  # noqa: E501
-            if self._execute(guest, package_manager_class.probe_command)
-        ]
+        ] = sorted(plugin_classes, key=lambda pm: pm.probe_priority, reverse=True)
 
-        discovered_package_managers.sort(key=lambda pm: pm.probe_priority, reverse=True)
-
-        if discovered_package_managers:
-            guest.debug(
-                'Discovered package managers',
-                fmf.utils.listed([pm.NAME for pm in discovered_package_managers]),
-                level=4,
-            )
-
-            return discovered_package_managers[0].NAME
+        for package_manager_class in discovered_package_managers:
+            if self._execute(guest, package_manager_class.probe_command):
+                guest.debug(
+                    f'Discovered {debug_label}',
+                    package_manager_class.NAME,
+                    level=4,
+                )
+                return package_manager_class.NAME
 
         return None
 
-    def _query_bootc_builder(
+    def _query_package_manager(
         self, guest: 'Guest'
     ) -> Optional['tmt.package_managers.GuestPackageManager']:
-        # Discover as many package managers as possible: sometimes, the
-        # first discovered package manager is not the only or the best
-        # one available. Collect them, and sort them by their priorities
-        # to find the most suitable one.
-
-        discovered_package_managers: list[
-            PackageManagerClass[tmt.package_managers.PackageManagerEngine]
-        ] = []
-
-        for (
-            package_manager_class
-        ) in tmt.package_managers._PACKAGE_MANAGER_PLUGIN_REGISTRY.iter_plugins():
-            if not package_manager_class.bootc_builder:
-                continue
-
-            if self._execute(guest, package_manager_class.probe_command):
-                discovered_package_managers.append(package_manager_class)
-
-        discovered_package_managers.sort(key=lambda pm: pm.probe_priority, reverse=True)
-
-        if discovered_package_managers:
-            guest.debug(
-                'Discovered bootc builders',
-                fmf.utils.listed([pm.NAME for pm in discovered_package_managers]),
-                level=4,
-            )
-
-            return discovered_package_managers[0].NAME
+        return self._discover_package_manager(
+            guest,
+            plugin_classes=[
```

Comment:
Since an interable is accepted, we should be able to sneak in a generator, `(package_manager_class for _, package_manager_class in (tmt.package_managers._PACKAGE_MANAGER_PLUGIN_REGISTRY.items())` instead of a materialized list.

Reply:
Hm, maybe the code was slightly different, like a `list()` call instead of actual generator. I would still give it a try.